plugin/plugins/filesystem/nodes/root.go

```diff
@@ -0,0 +1,216 @@
+var _ walkRef = (*RootIndex)(nil)
+
+const nRoot = "" // root namespace is intentionally left blank
+
+type rootPath string
+
+func (rp rootPath) String() string { return string(rp) }
+func (rootPath) Namespace() string { return nRoot }
+func (rootPath) Mutable() bool     { return true }
+func (rootPath) IsValid() error    { return nil }
+func (rootPath) Root() cid.Cid     { return cid.Cid{} }
+func (rootPath) Remainder() string { return "" }
+func (rp rootPath) Cid() cid.Cid {
+	prefix := cid.V1Builder{Codec: cid.DagCBOR, MhType: multihash.BLAKE2B_MIN}
+	c, err := prefix.Sum([]byte(rp))
+	if err != nil {
+		panic(err) //invalid root
+	}
+	return c
+}
+
+type systemTuple struct {
+	file   p9.File
+	dirent p9.Dirent
+}
+
+type systemSlice []systemTuple
+
+func (ss systemSlice) Len() int           { return len(ss) }
+func (ss systemSlice) Swap(i, j int)      { ss[i], ss[j] = ss[j], ss[i] }
+func (ss systemSlice) Less(i, j int) bool { return ss[i].dirent.Offset < ss[j].dirent.Offset }
+
+//TODO: rename, while this is likely to be the root, it doesn't have to be; maybe "IPFSOverlay"
+// RootIndex is a virtual directory file system, that maps a set of file system implementations to a hierarchy
+// Currently: "/ipfs":PinFS, "/ipfs/*:IPFS
+type RootIndex struct {
+	Base
+	//subsystems []p9.Dirent
+	subsystems map[string]systemTuple
+	core       coreiface.CoreAPI
+}
+
+// RootAttacher constructs the default RootIndex file system, providing a means to Attach() to it
+func RootAttacher(ctx context.Context, core coreiface.CoreAPI, ops ...nodeopts.Option) p9.Attacher {
+	ri := &RootIndex{
+		core: core,
+		Base: Base{
+			Logger:    logging.Logger("RootFS"),
+			parentCtx: ctx,
+			Qid: p9.QID{
+				Type: p9.TypeDir,
+				Path: cidToQPath(rootPath("/").Cid()),
+			},
+		},
+	}
+	ri.meta, ri.metaMask = defaultRootAttr()
+
+	rootDirent := p9.Dirent{
+		Type: p9.TypeDir,
+		QID:  p9.QID{Type: p9.TypeDir},
+	}
+
+	type subattacher func(context.Context, coreiface.CoreAPI, ...nodeopts.Option) p9.Attacher
+	type attachTuple struct {
+		string
+		subattacher
+	}
+	subsystems := [...]attachTuple{
+		{"ipfs", PinFSAttacher},
+	}
+
+	ri.subsystems = make(map[string]systemTuple, len(subsystems))
+	opts := []nodeopts.Option{nodeopts.Parent(ri)}
+
+	for i, subsystem := range subsystems {
+		fs, err := subsystem.subattacher(ctx, core, opts...).Attach()
+		if err != nil {
+			panic(err) // hard implementation error
+		}
+
+		rootDirent.Offset = uint64(i + 1)
+		rootDirent.Name = subsystem.string
+
+		rootDirent.QID.Path = cidToQPath(rootPath("/" + subsystem.string).Cid())
+
+		ri.subsystems[subsystem.string] = systemTuple{
+			file:   fs,
+			dirent: rootDirent,
+		}
+	}
+
+	options := nodeopts.AttachOps(ops...)
+	if options.Parent != nil {
+		ri.parent = options.Parent
+	} else {
+		ri.parent = ri
+		ri.root = true
+	}
+
+	return ri
+}
+
+func (ri *RootIndex) Attach() (p9.File, error) {
+	ri.Logger.Debugf("Attach")
+
+	newFid := new(RootIndex)
+	*newFid = *ri
```

Comment:
I really don't think we want to _copy_ everything here, right? If we need to keep shared state, we may want to break this out into a separate object.

Reply:
Behaviour around this has changed. Instead of using memory copies directly, we now have 2 constructs "clone" and "fork" which are used internally by the FS.
Clone makes a near direct copy, ommitting only certain fields such as context cancel function pointers. The reference returned is effectively the same, but may be closed independently.
Fork is similar to Clone but it is implied that the returned reference is independent, holding its own context, cancel, path, etc.